File: java/strscan/src/test/StrScannerTest.java
package test;

import static org.junit.Assert.assertEquals;

import org.junit.Test;
import org.junit.runner.RunWith;
import de.bechte.junit.runners.context.*;
import src.StrScanner;

@RunWith(HierarchicalContextRunner.class)
public class StrScannerTest {
	@Test
	public void isInitializedWithAString() {
		String string = "abc def";
		String expected = string;
		
		StrScanner scanner = new StrScanner(string);
		String actual = scanner.getString();
				
		assertEquals(expected, actual);
	}
	
	public class Eof {
		@Test
		public void returnsFalseIfNotAtEndOfString() {
			String string = "abc def";
			boolean expected = false;
			
			StrScanner scanner = new StrScanner(string);
			boolean actual = scanner.eof();
					
			assertEquals(expected, actual);
		}
		
		@Test
		public void returnsTrueIfAtEndOfString() {
			String string = "";
			boolean expected = true;
			
			StrScanner scanner = new StrScanner(string);
			boolean actual = scanner.eof();
					
			assertEquals(expected, actual);
		}
	}
	
	public class Scan {
		@Test
		public void returnsMatchIfFound() {
			String string = "abc";
			String regex = "abc";
			
			String expected = "abc";
			
			StrScanner scanner = new StrScanner(string);
			String actual = scanner.scan(regex);
					
			assertEquals(expected, actual);
		}
		
		@Test
		public void returnsNextMatch() {
			String string = "abc def";
			String regex = "abc";
			
			String expected = "abc";
			
			StrScanner scanner = new StrScanner(string);
			String actual = scanner.scan(regex);
					
			assertEquals(expected, actual);
		}
		
		@Test
		public void returnsNullIfMatchIsntFirst() {
			String string = "abc def";
			String regex = "\\s+";
			
			String expected = null;
			
			StrScanner scanner = new StrScanner(string);
			String actual = scanner.scan(regex);
					
			assertEquals(expected, actual);
		}
		
		@Test
		public void removesMatchFromString() {
			String string = "abc def";
			String regex = "\\w+";
			
			String expected = " def";
			
			StrScanner scanner = new StrScanner(string);
			scanner.scan(regex);
			String actual = scanner.getString();
					
			assertEquals(expected, actual);
		}
	}
}

File: elixir/strscan/test/str_scanner_test.exs
defmodule StrScannerTest do
  use ExUnit.Case
  doctest StrScan

  test "eof?/1 returns false if string is not empty" do
    string = "abc def"
    expected = false

    actual = StrScan.eof?(string)

    assert expected == actual
  end

  test "eof?/1 returns true if string is empty" do
    string = ""
    expected = true

    actual = StrScan.eof?(string)

    assert expected == actual
  end

  test "scan/2 returns match and remainder regex" do
    string = "abc"
    regex = ~r{abc}
    expected = ["abc", ""]

    actual = StrScan.scan(string, regex)

    assert expected == actual
  end

  test "scan/2 returns match and remainder regex if string starts with match" do
    string = "abc def"
    regex = ~r{abc}
    expected = ["abc", " def"]

    actual = StrScan.scan(string, regex)

    assert expected == actual
  end

  test "scan/2 returns nil and string if string does not start with match" do
    string = "abc def"
    regex = ~r{\s+}
    expected = [nil, "abc def"]

    actual = StrScan.scan(string, regex)

    assert expected == actual
  end
end
